### ISSUE TITLE
How to build forms in DSpace 7

### ISSUE THREAD

**Issue by atarix83** [DSpace/dspace-angular#133]
For the submission functionalities it's necessary to implement a component that allows to generate edit forms dynamically.

Should it be reasonable to use third-party library like [https://github.com/udos86/ng2-dynamic-forms](https://github.com/udos86/ng2-dynamic-forms)?

It looks to have some very useful functionalities and allows to use NG Bootstrap as UI Library.

I like to hear other opinion on that.

Thanks
Giuseppe

**Comment by tdonohue** [comment]
@atarix83 : It seems like we have a few decision points we'd need to consider before selecting an angular plugin for forms. Ideally, we'd want the same plugin to work for simple Edit Item/Community/Collection Forms as well as for more dynamic Submission Forms

- How will Submission form field information be made available to the Angular UI?  Presumably, the REST API will need an endpoint for Submission form configuration (pulled from `item-submission.xml` and `input-forms.xml` config files).
- Does the plugin support all the various form controls we require (for either edit or submissions)?  I notice that `n2-dynamic-forms` has a "compatibility table" in the [UI Components](https://github.com/udos86/ng2-dynamic-forms#ui-components) section. It notes that, based on the UI package, certain form controls are not yet supported.

To be honest, the `ng2-dynamic-forms` plugin looks reasonable at a glance, but I have no experience with it. So, we'd want to be sure it'll meet the needs for all forms we plan to create in the new UI.

**Comment by artlowel** [comment]
ng2-dynamic-forms looks pretty comprehensive and well maintained, however I don't find any information about using it with ngrx. 
That is something we'd need to test before we can commit to using it.
To answer questions like:
- When an input's value changes, can we get it to dispatch an ngrx action instead of updating a model local to the component?
- Can we use observables on the store to represent values in the forms?
- Can we use observables on the store to represent the form configuration?
- What about the status of a form field? whether it has been modified, validated, etc. Can the state of all these things be kept in the ngrx store?

**Comment by atarix83** [comment]
yes we need to test it before.

about using with ngrx maybe [this example](https://github.com/angular-redux/form/tree/master/source) should be a good point to borrow a few ideas from

**Comment by atarix83** [comment]
Hi,

[Here](https://github.com/4Science/dspace-angular/tree/forms) a proof of concept where I've tried to use ngrx with ng2-dynamic-forms.

I got started from [this project](https://github.com/angular-redux/form) and I have adapted it to works with ngrx.

It's just a draft and I need to reengineer it whether we decide to follow this way.

Let me know what do you think it.

Thanks

**Comment by artlowel** [comment]
It seems that the state is saved in the store, so that's good, but the forms don't update if the store updates. You can test this by manually dispatching a FORM_CHANGED action from the redux devtools ([example](https://www.dropbox.com/s/ajuguzagugdj5b7/Sep-01-2017%2016-11-59.mp4?dl=0)). If you change a field in the state, its corresponding form field should change as well.

I'm also worried about the amount of code that is copied from that project you mentioned and dropped in to dspace. It would be different if this was a dependency that could be added to package.json, but in this case we're simply copying and pasting a bunch of quite complex code. That won't make it easy to ever upgrade, or bugfix.

Finally that code also makes use of immutable.js. We agreed in [the meeting of march 2nd](https://wiki.duraspace.org/display/DSPACE/2017-03-02+DSpace+7+UI+Working+Group+Meeting+notes) not to use it, because while it makes writing your reducers a little easier, it can also cause performance issues, makes things harder to debug, etc.

**Comment by atarix83** [comment]
thanks for the review.

I agree with you about importing code is not a good solution, It should be better to fork that project and create a new package maybe.

In consideration of this and the use of immutable.js, may it be preferable to build our wrapper with a new ngrx/store implementation?

**Comment by atarix83** [comment]
@tdonohue @artlowel 

[Here](https://github.com/4science/dspace-angular/tree/forms) the first implementation for the form builder used in the submission module.

Essentially I chose to use ngx-dynamic-forms core module, and to build our DSpace UI library based on NG Bootstrap UI Library. This is because we need more type of form fields that can't be rendered only using NG Bootstrap as UI Library.

Form's State implementation should be meet the requirements. 

To test it go to the url http://localhost:3000/forms.
The form you can view in that page is built by a mock input-form definition, that is similar of the object that it will be retrieve from new rest endpoint.

Some fields type must be implemented, but you can take a look to 'Other Titles' and 'Types' fields, which have been realized by DSpace UI library.

Let me know what do you think about it.

Thanks

**Comment by artlowel** [comment]
It works. I tried to change a few things to the configuration and they showed up in the form. The reducer and assorted classes look good and modify the data in an immutable way.

However when I see files with huge switches like [ds-dynamic-form-control.component.html](https://github.com/4Science/dspace-angular/blob/ea72607e9783bba555b72f458fb7385492efdc41/src/app/shared/form/builder/ds-dynamic-form-ui/ds-dynamic-form-control.component.html) or [form-builder.service.ts](https://github.com/4Science/dspace-angular/blob/ea72607e9783bba555b72f458fb7385492efdc41/src/app/shared/form/builder/form-builder.service.ts) I get worried.

For the service, I understand it parses the form definitions that come from the backend. I would have expected most of the heavy lifting to be done by [cerealize](https://github.com/weichx/cerialize). The lib we use to turn the json representations of the other DSpaceObjects in to models. It allows you to map the model's properties to the response using annotations.

And for the template: Lotte had a similar issue where she wanted to make a single component to render a different subcomponent depending on what type of DSO the model was (Item, Collection, …). She ended up writing a custom annotation `@listElementFor`. As a consequence the template that would have contained the switch can look like [this](https://github.com/DSpace/dspace-angular/blob/563cf6e8200df3a34db31c53fb62c0aad39dbccd/src/app/object-list/wrapper-list-element/wrapper-list-element.component.html), and the sub template for each DSO [gets that annotation](https://github.com/DSpace/dspace-angular/blob/dd2b8b76d949acd4ba93c253dd79b65d544b7fb6/src/app/object-list/collection-list-element/collection-list-element.component.ts#L13) to specify which model it should render. Perhaps a similar technique could be used here?

**Comment by atarix83** [comment]
@artlowel 

About the template solution, it's a good hint, I can try to apply here.

Meanwhile about the use of cerealize, the main problem is that most of the model used are include in the Core library of ng-dynamic-forms and to apply meta data annotations I should override them. Furthermore the form definitions is already deserialized when is retrieved from the backend

**Comment by artlowel** [comment]
>Furthermore the form definitions is already deserialized when is retrieved from the backend

Then why do you need a [modelFromConfiguration](https://github.com/4Science/dspace-angular/blob/ea72607e9783bba555b72f458fb7385492efdc41/src/app/shared/form/builder/form-builder.service.ts#L39) method? That seems to be exactly the kind of thing cerialize was made for.

Even if you don't end up using cerialize I'd still appreciate it if you broke up those methods in to smaller parts. A 150+ line method is just bad practice. 

e.g. something like: 
```typescript
      switch (fieldData.input.type) {
        case 'date':
          group.push(new DateFieldParser(fieldData).parse());
          break;

        case 'dropdown':
          group.push(new DropdownFieldParser(fieldData).parse());
          break;
        ...
```
should be possible, and would already be a big improvement in my opinion. Those FieldParsers could 
possibly extend a common ancestor and share some code, etc.

**Comment by tdonohue** [comment]
Closing as fixed by #264
